### PR TITLE
[CFF] Fix dict values so can be addressed from a null pool

### PR DESCRIPTION
**Pull request by ebraminio**
fixes https://crbug.com/971933

### CODE REVIEW

File: src/hb-ot-cff1-table.hh [hunk 1]
```diff
@@ -895,7 +895,14 @@ struct cff1_private_dict_values_base_t : dict_values_t<VAL>
     return size;
   }
 
+  const CFF1Subrs* get_subrs () const
+  { return localSubrs ? localSubrs : &Null (CFF1Subrs); }
+
+  void set_subrs (const CFF1Subrs* _localSubrs)
+  { localSubrs = _localSubrs; }
+
   unsigned int      subrsOffset;
```
Comment (ebraminio) [inline src/hb-ot-cff1-table.hh]: This also should become private but wanted to do that in separate change

**Comment by behdad** [comment]
Use `hb_nonnull_ptr_t`?  I don't know the CFF code well, so cannot really review.

**Comment by ebraminio** [comment]
> Use hb_nonnull_ptr_t? I don't know the CFF code well, so cannot really review.

Done

File: src/hb-subset-cff-common.hh [hunk 2]
```diff
@@ -658,9 +658,8 @@ struct subr_subsetter_t
     parsed_global_subrs.init (acc.globalSubrs->count);
     parsed_local_subrs.resize (acc.fdCount);
     for (unsigned int i = 0; i < acc.fdCount; i++)
-    {
-      parsed_local_subrs[i].init (acc.privateDicts[i].localSubrs->count);
-    }
+      parsed_local_subrs[i].init (acc.privateDicts[i].localSubrs.get ()->count);
```
Comment (behdad) [inline src/hb-subset-cff-common.hh]: Don't actually need `.get()` before dereferencing...

File: src/hb-ot-cff2-table.hh [hunk 3]
```diff
@@ -491,8 +492,8 @@ struct cff2
 	if (unlikely (!priv_interp.interpret (privateDicts[i]))) { fini (); return; }
 
 	privateDicts[i].localSubrs = &StructAtOffsetOrNull<CFF2Subrs> (&privDictStr[0], privateDicts[i].subrsOffset);
-	if (privateDicts[i].localSubrs != &Null(CFF2Subrs) &&
-	  unlikely (!privateDicts[i].localSubrs->sanitize (&sc)))
+	if (privateDicts[i].localSubrs.get () != &Null (CFF2Subrs) &&
```
Comment (behdad) [inline src/hb-ot-cff2-table.hh]: Maybe add is_null() to hb_nonnull_ptr_t?

The way these things were intended to be used, one should never compare to null object...

**Comment by blueshade7** [comment]
The suggested changes look ok, but they don’t seem to make the original problem go away. I still get the error when I run shape fuzzer with the test case.

Complaint appears to come from the fact that a null Subrs pointer is derefed (*) then passed as a ref param (&) to `cs_interp_env_t::init` then to `biased_subrs_t::init`, even though they are actually used as pointers after refed (&). My bad for this inconsistency. So my suggestion is to change these init functions to take Subrs params as pointers instead of references.

Shall I push this change to cffnull branch?

File: src/hb-ot-cff2-table.hh (hunk 3 above)
Comment (ebraminio) [reply src/hb-ot-cff2-table.hh]: Yep, I don't want put comparison to null pool there so we should compare to nullptr itself and that means removing the above line, which, has unintended effect so lets skip this for now.

**Comment by ebraminio** [comment]
> The suggested changes look ok, but they don’t seem to make the original problem go away. I still get the error when I run shape fuzzer with the test case.

Are you sure testing the branch? I don't see the issue both locally and in the bots and they were complaining when I first uploaded the testcase without the fix.

> Complaint appears to come from the fact that a null Subrs pointer is derefed (*) then passed as a ref param (&) to cs_interp_env_t::init then to biased_subrs_t::init, even though they are actually used as pointers after refed (&). My bad for this inconsistency. So my suggestion is to change these init functions to take Subrs params as pointers instead of references.

I believe complaint is coming from `*acc.privateDicts[fd].localSubrs` that fd is over the `privateDicts` `hb_vector_t` size which means the returned values from indexing will be from null pool and this fixes it by making it not need to explicitly `init ()`ing needed so the returned value from null pool can work also.

> Shall I push this change to cffnull branch?

Feel free to merge this first, and/or, create a separate PR based on these changes or not, I am pretty sure this is the fix the testcase, I should've explained why.

**Comment by blueshade7** [comment]
> Are you sure testing the branch?

Ah I failed to grab branch from your depot. As the result I accidentally tested on master and came up with another fix as outlined above. Anyways I created a branch for it as `harfbuzz/cffnull_fix2`

**Comment by ebraminio** [comment]
Great, please upload also as a PR and put the testcase on the patch. Thanks

File: src/hb-ot-cff1-table.hh [hunk 4]
```diff
@@ -1138,8 +1137,7 @@ struct cff1
 	if (unlikely (!priv_interp.interpret (*priv))) { fini (); return; }
 
 	priv->localSubrs = &StructAtOffsetOrNull<CFF1Subrs> (&privDictStr, priv->subrsOffset);
-	if (priv->localSubrs != &Null(CFF1Subrs) &&
-	    unlikely (!priv->localSubrs->sanitize (&sc)))
+	if (priv->subrsOffset && !priv->localSubrs->sanitize (&sc))
```
Comment (ebraminio) [inline src/hb-ot-cff1-table.hh]: Apparently CFF1Subrs::sanitize (and CFF2Subrs) isn't returning true on NullPool objects which isn't good.

Comment (ebraminio) [reply src/hb-ot-cff1-table.hh]: `Number of object data. Note there are (count+1) offsets` hmm, so we can't do anything about

**Comment by ebraminio** [comment]
Lets deal with this later

**Comment by behdad** [comment]
Just commit whitespace changes separately and push to master... or not make whitespace changes.

**Pull request closed without merging**